apis/Google.Longrunning/Google.Longrunning/Operation.cs

```diff
@@ -0,0 +1,329 @@
+﻿// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Api.Gax;
+using Google.Protobuf;
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Google.Longrunning
+{
+    /// <summary>
+    /// A long-running operation with an associated client, and which knows the expected response type.
+    /// </summary>
+    /// <remarks>
+    /// For simplicity, no methods on this type modify the proto message. Instead, to get up-to-date
+    /// information you can use Refresh to obtain a new instance.
+    /// </remarks>
+    /// <typeparam name="T">The response message type.</typeparam>
+    public class Operation<T> where T : IMessage<T>, new()
+    {
+        /// <summary>
+        /// The protobuf message associated with the long-running operation, containing the name (for
+        /// further retrieval) and any error/result already computed. This should not be mutated.
+        /// </summary>
+        public Operation RpcMessage { get; }
+
+        /// <summary>
+        /// The client to use when making RPCs.
+        /// </summary>
+        public OperationsClient Client { get; }
+
+        private readonly Lazy<OperationFailedException> _lazyException;
+
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <param name="rpcMessage"></param>
+        /// <param name="client"></param>
+        public Operation(Operation rpcMessage, OperationsClient client)
+        {
+            RpcMessage = GaxPreconditions.CheckNotNull(rpcMessage, nameof(rpcMessage));
+            Client = GaxPreconditions.CheckNotNull(client, nameof(rpcMessage));
+            _lazyException = new Lazy<OperationFailedException>(
+                () => RpcMessage.Error == null ? null : new OperationFailedException(RpcMessage),
+                LazyThreadSafetyMode.ExecutionAndPublication);
```

Comment:
Nice to know - but I think I'd rather leave it explicit here so that the reader doesn't need to check that :)